task/mgmt-gateway/src/update/sp/auxflash.rs

```diff
@@ -0,0 +1,335 @@
+                        total: SLOT_COUNT + SECTORS_PER_SLOT as u32,
+                    }),
+                })
+            }
+            Self::ErasingSlot(erase) => {
+                UpdateStatus::Preparing(UpdatePreparationStatus {
+                    id,
+                    progress: Some(UpdatePreparationProgress {
+                        current: SLOT_COUNT + erase.sectors_to_erase.start,
+                        total: SLOT_COUNT + SECTORS_PER_SLOT as u32,
+                    }),
+                })
+            }
+            Self::FinishedErasingSlot(_) => {
+                UpdateStatus::SpUpdateAuxFlashChckScan {
+                    id,
+                    found_match: false,
+                    total_size,
+                }
+            }
+            Self::AcceptingData(data) => {
+                UpdateStatus::InProgress(UpdateInProgressStatus {
+                    id,
+                    bytes_received: data.next_write_offset
+                        + data.buffer.len() as u32,
+                    total_size,
+                })
+            }
+            Self::Failed(err) => {
+                UpdateStatus::Failed {
+                    id,
+                    // TODO These error codes conflict with `UpdateError` codes
+                    // if we failed during the SP image update; is that okay?
+                    // May be clear from context which code it is.
+                    code: *err as u32,
+                }
+            }
+        }
+    }
+}
+
+pub(super) struct ScanningForChck {
+    buffer: UpdateBuffer,
+    chck: AuxFlashChecksum,
+    // The currently-active slot for our running image. This is the slot
+    // index where we start scanning, since we typically expect to find a
+    // match to our current aux flash image. If we have no active slot
+    // (i.e., this is `None`), it is treated as 0.
+    active_slot: Option<u32>,
+    // Index of the slot to scan next. `active_slot - 1` (after wrapping
+    // around at SLOT_COUNT) is the final slot to scan.
+    index: u32,
+    // While we're scanning, we record the first empty, even-numbered slot
+    // we see. If we don't find a CHCK match, we'll pick this slot to
+    // erase/write. If we don't find an empty, even slot, we'll the next
+    // even slot above our current active slot.
+    first_empty_even_slot: Option<u32>,
+}
+
+impl ScanningForChck {
+    pub(super) fn continue_scanning(
+        mut self,
+        task: &AuxFlash,
+    ) -> ChckScanResult {
+        // Scan the slot at `index`.
+        match task.read_slot_chck(self.index) {
+            Ok(chck) => {
+                // If this matches, we're done; transition to the
+                // next state and return the buffer to our caller.
+                if chck == self.chck {
+                    return ChckScanResult::FoundMatch(self.buffer);
+                }
+            }
+            Err(AuxFlashError::MissingChck) => {
```

Comment:
I think that `BadChckSize`, `MissingAuxi`, `MultipleChck`, `MultipleAuxi`, and `ChckMismatch` should all be in this section: treating corrupt slots as empty seems reasonable, since they're useless otherwise.

Reply:
:+1: added in e966a578